File: day07/js/solution.js
import fs from "fs";

const grid = fs.readFileSync("../input.txt", "utf-8").trimEnd().split("\n");

function part1(grid) {
  const rows = grid.length;
  const cols = grid[0].length;

  const startCol = grid[0].indexOf("S");

  let active = new Set([startCol]);
  let splits = 0;

  for (let r = 1; r < rows; r++) {
    const nextActive = new Set();

    for (const c of active) {
      if (c < 0 || c >= cols) continue;

      const cell = grid[r][c];

      if (cell === ".") {
        nextActive.add(c);
      } else if (cell === "^") {
        splits++;
        nextActive.add(c - 1);
        nextActive.add(c + 1);
      }
    }

    active = nextActive;
  }

  return splits;
}

function part2(grid) {
  const rows = grid.length;
  const cols = grid[0].length;

  const startCol = grid[0].indexOf("S");

  let current = new Map();
  current.set(startCol, 1);

  for (let r = 1; r < rows; r++) {
    const next = new Map();

    for (const [c, count] of current.entries()) {
      if (c < 0 || c >= cols) continue;

      const cell = grid[r][c];

      if (cell === ".") {
        next.set(c, (next.get(c) ?? 0) + count);
      } else if (cell === "^") {
        next.set(c - 1, (next.get(c - 1) ?? 0) + count);
        next.set(c + 1, (next.get(c + 1) ?? 0) + count);
      }
    }

    current = next;
  }

  let total = 0;
  for (const v of current.values()) {
    total += v;
  }

  return total;
}

console.log("Part 1:", part1(grid));
console.log("Part 2:", part2(grid));

File: day07/py/solution.py
from pathlib import Path
from collections import defaultdict

grid = Path("../input.txt").read_text().splitlines()

def part1(grid):
  rows = len(grid)
  cols = len(grid[0])

  start_col = grid[0].index("S")

  active = {start_col}
  splits = 0

  for r in range(1, rows):
      next_active = set()

      for c in active:
          if c < 0 or c >= cols:
              continue

          cell = grid[r][c]

          if cell == ".":
              next_active.add(c)

          elif cell == "^":
              splits += 1
              next_active.add(c - 1)
              next_active.add(c + 1)

      active = next_active

  return splits


def part2(grid):
    rows = len(grid)
    cols = len(grid[0])

    start_col = grid[0].index("S")

    current = {start_col: 1}

    for r in range(1, rows):
        next_state = defaultdict(int)

        for c, count in current.items():
            if c < 0 or c >= cols:
                continue

            cell = grid[r][c]

            if cell == ".":
                next_state[c] += count

            elif cell == "^":
                next_state[c - 1] += count
                next_state[c + 1] += count

        current = next_state

    total = sum(current.values())

    return total

print("Part 1:", part1(grid))
print("Part 2:", part2(grid))
